src/app/fdctl/run/tiles/fd_net.c

```
@@ -303,31 +215,92 @@ struct xdp_statistics_v1 {
+    stats.rx_invalid_descs         += sub_stats.rx_invalid_descs;
+    stats.tx_invalid_descs         += sub_stats.tx_invalid_descs;
+    stats.rx_ring_full             += sub_stats.rx_ring_full;
+    stats.rx_fill_ring_empty_descs += sub_stats.rx_fill_ring_empty_descs;
+    stats.tx_ring_empty_descs      += sub_stats.tx_ring_empty_descs;
   }
+
+  FD_MCNT_SET( NET, XDP_RX_DROPPED_OTHER,         stats.rx_dropped               );
+  FD_MCNT_SET( NET, XDP_RX_INVALID_DESCS,         stats.rx_invalid_descs         );
+  FD_MCNT_SET( NET, XDP_TX_INVALID_DESCS,         stats.tx_invalid_descs         );
+  FD_MCNT_SET( NET, XDP_RX_RING_FULL,             stats.rx_ring_full             );
+  FD_MCNT_SET( NET, XDP_RX_FILL_RING_EMPTY_DESCS, stats.rx_fill_ring_empty_descs );
+  FD_MCNT_SET( NET, XDP_TX_RING_EMPTY_DESCS,      stats.tx_ring_empty_descs      );
+}
+
+/* net_tx_ready returns 1 if the current XSK is ready to submit a TX send
+   job.  If the XSK is blocked for sends, returns 0.  Reasons for block
+   include:
+   - No XSK TX buffer is available
+   - XSK TX ring is full */
+
+static int
+net_tx_ready( fd_net_ctx_t * ctx,
+              uint           if_idx ) {
+  fd_xsk_t *           xsk     = ctx->xsk[ if_idx ];
+  fd_ring_desc_t *     tx_ring = &xsk->ring_tx;
+  fd_net_free_ring_t * free    = ctx->free_tx + if_idx;
+  if( free->prod == free->cons ) return 0; /* drop */
+  if( tx_ring->prod - tx_ring->cons >= tx_ring->depth ) return 0; /* drop */
+  return 1;
+}
+
+/* net_rx_wakeup triggers xsk_recvmsg to run in the kernel.  Needs to be
+   called periodically in order to receive packets. */
+
+static void
+net_rx_wakeup( fd_net_ctx_t * ctx,
+               fd_xsk_t *     xsk ) {
+  (void)ctx; /* FIXME metrics */
+  if( !fd_xsk_rx_need_wakeup( xsk ) ) return;
+  struct msghdr _ignored[ 1 ] = { 0 };
+  if( FD_UNLIKELY( -1==recvmsg( xsk->xsk_fd, _ignored, MSG_DONTWAIT ) ) ) {
+    if( FD_UNLIKELY( errno!=EAGAIN ) ) {
```

I really really hate this swallowing of fatal programmer errors. What are the actual expected / possible errnos here? Can we just WARN on those and terminate on the others? I presume we do not want to keep running the app if we get EACCESS here for example